### PR TITLE
NSM: TFTP eve logging with rust

### PR DESCRIPTION
**Pull request by macpas**
Make sure these boxes are signed before submitting your Pull Request -- thank you.
Update of PR 3113
TFTP parsing and logging written in Rust.
Log on eve.json the type of request (read or write), the name of the file and the mode.
(with tftp tag is added in your eve-log yaml section).

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- ability to log tftp record in eve json
- available only with rust activated

example:
"event_type":"tftp","src_ip":"192.168.0.253","src_port":50618,"dest_ip":"192.168.0.10","dest_port":69,"proto":"UDP","tftp":{"packet":"read","file":"rfc1350.txt","mode":"octet"}
"event_type":"tftp","src_ip":"192.168.0.1","src_port":57509,"dest_ip":"192.168.0.13","dest_port":69,"proto":"UDP","tftp":{"packet":"write","file":"rfc1350.txt","mode":"octet"}

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

File: suricata.yaml.in [hunk 1]
```diff
@@ -238,6 +238,7 @@ outputs:
             #md5: [body, subject]
 
         #- dnp3
+        #- tftp
```
Comment (victorjulien) [inline suricata.yaml.in]: should this be enabled by default if rust is built in? So using @rust_config_comment@

Comment (victorjulien) [reply suricata.yaml.in]: Seems a app-layer config option is missing

File: src/app-layer-tftp.c [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+/**
+ * \brief Probe the input to see if it looks like echo.
+ *
+ * \retval ALPROTO_TFTP if it looks like echo, otherwise
+ *     ALPROTO_UNKNOWN.
+ */
+static AppProto TFTPProbingParser(Flow *f, uint8_t *input, uint32_t input_len,
+    uint32_t *offset)
+{
+    /* Very simple test - if there is input, this is echo.
```
Comment (victorjulien) [inline src/app-layer-tftp.c]: 'this is echo' -> tftp

File: rust/src/tftp/tftp.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        return std::ptr::null_mut();
+    }
+    return unsafe{transmute(&state.transactions[tx_id as usize])};
+}
+
+#[no_mangle]
+pub extern "C" fn rs_tftp_get_tx_logged(_state: &mut TFTPState,
+                                        tx: &mut TFTPTransaction,
+                                        logger: libc::uint32_t) -> i8 {
+    if tx.logged.is_logged(logger) {
```
Comment (victorjulien) [inline rust/src/tftp/tftp.rs]: this is now outdated, see https://github.com/OISF/suricata/commit/ec62eedc87b13f3f5dbcd845c46b601daed3ca93. Fix is simple, see https://github.com/OISF/suricata/commit/5c6868b327dda74b4f0bb7934923a4a93a2910c7

File: src/output-json-tftp.c [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+/**
+ * \file
+ *
+ * \author Clément Galland <clement.galland@epita.fr>
+ *
+ * Implement JSON/eve logging app-layer TFTP.
+ */
+
+//#ifdef HAVE_RUST
```
Comment (victorjulien) [inline src/output-json-tftp.c]: minor: remove

**Comment by macpas** [comment]
Hello,
attached public pcaps to this email.
Thanks

----- Mail original -----
De: "Victor Julien" <notifications@github.com>
À: "OISF/suricata" <suricata@noreply.github.com>
Cc: "macpas" <pdl35@free.fr>, "Author" <author@noreply.github.com>
Envoyé: Mardi 16 Janvier 2018 15:14:01
Objet: Re: [OISF/suricata] NSM: TFTP eve logging with rust (#3120)



@victorjulien requested changes on this pull request. 

Looks good w/o testing. Some minor comments & requesting another rebase. Do you have a public pcap that we can add to our test suite? 


In suricata.yaml.in : > @@ -238,6 +238,7 @@ outputs:
             #md5: [body, subject]
 
         #- dnp3
+        #- tftp 

should this be enabled by default if rust is built in? So using @rust_config_comment@ 


In suricata.yaml.in : > @@ -238,6 +238,7 @@ outputs:
             #md5: [body, subject]
 
         #- dnp3
+        #- tftp 

Seems a app-layer config option is missing 


In src/app-layer-tftp.c : > +
+static int TFTPHasEvents(void *state)
+{
+    return rs_tftp_has_event(state);
+}
+
+/**
+ * \brief Probe the input to see if it looks like echo.
+ *
+ * \retval ALPROTO_TFTP if it looks like echo, otherwise
+ *     ALPROTO_UNKNOWN.
+ */
+static AppProto TFTPProbingParser(Flow *f, uint8_t *input, uint32_t input_len,
+    uint32_t *offset)
+{
+    /* Very simple test - if there is input, this is echo. 

'this is echo' -> tftp 


In rust/src/tftp/tftp.rs : > +}
+
+#[no_mangle]
+pub extern "C" fn rs_tftp_get_tx(state: &mut TFTPState,
+                                    tx_id: libc::uint64_t) -> *mut libc::c_void {
+    if state.transactions.len() <= tx_id as usize {
+        return std::ptr::null_mut();
+    }
+    return unsafe{transmute(&state.transactions[tx_id as usize])};
+}
+
+#[no_mangle]
+pub extern "C" fn rs_tftp_get_tx_logged(_state: &mut TFTPState,
+                                        tx: &mut TFTPTransaction,
+                                        logger: libc::uint32_t) -> i8 {
+    if tx.logged.is_logged(logger) { 

this is now outdated, see ec62eed . Fix is simple, see 5c6868b 


In src/output-json-tftp.c : > + *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Clément Galland <clement.galland@epita.fr>
+ *
+ * Implement JSON/eve logging app-layer TFTP.
+ */
+
+//#ifdef HAVE_RUST 

minor: remove

File: rust/src/tftp/tftp.rs (hunk 3 above)
Comment (macpas) [reply rust/src/tftp/tftp.rs]: OK understood.

File: suricata.yaml.in (hunk 1 above)
Comment (macpas) [reply suricata.yaml.in]: First comment is understood (I will add the @rust_config_comment@- tftp).
I do not understand your last comment. Thanks.

**Comment by macpas** [comment]
replaced now by PR: 3176

**Pull request closed without merging**